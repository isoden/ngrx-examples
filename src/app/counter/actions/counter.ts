import { Action } from '@ngrx/store';

export enum CounterActionTypes {
  Increment = '[Counter] Increment',
  Decrement = '[Counter] Decrement',
}

export class Increment implements Action {
  readonly type = CounterActionTypes.Increment;
}

export class Decrement implements Action {
  readonly type = CounterActionTypes.Decrement;
}

export type CounterActions =
  | Increment
  | Decrement;
