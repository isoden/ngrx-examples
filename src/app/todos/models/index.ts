export { Todo } from './todo';
export { VisibilityFilter } from './visibility-filter';
