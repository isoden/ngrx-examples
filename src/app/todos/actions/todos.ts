import { Action } from '@ngrx/store';
import { VisibilityFilter } from '../models';

export enum TodosActionTypes {
  Add                 = '[Todos] Add',
  Toggle              = '[Todos] Toggle',
  SetVisibilityFilter = '[Todos] Set Visibility Filter',
}

export class Add implements Action {
  readonly type = TodosActionTypes.Add;

  constructor(public payload: { title: string }) {}
}

export class Toggle implements Action {
  readonly type = TodosActionTypes.Toggle;

  constructor(public payload: { id: number }) {}
}

export class SetVisibilityFilter implements Action {
  readonly type = TodosActionTypes.SetVisibilityFilter;

  constructor(public payload: { visibilityFilter: VisibilityFilter }) {}
}

export type TodosActions =
  | Add
  | Toggle
  | SetVisibilityFilter;
