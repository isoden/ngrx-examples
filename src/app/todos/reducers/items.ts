import { createEntityAdapter } from '@ngrx/entity';
import { Todo } from '../models';
import { EntityState } from '@ngrx/entity/src/models';
import { TodosActions, TodosActionTypes } from '../actions/todos';

export interface TodosState extends EntityState<Todo> {
  lastId: number;
}

export const adapter = createEntityAdapter<Todo>();
const initialState = adapter.getInitialState<Pick<TodosState, 'lastId'>>({
  lastId: 0,
});

export function reducer(state = initialState, action: TodosActions): TodosState {
  switch (action.type) {
    case TodosActionTypes.Add: {
      const todo = new Todo(state.lastId + 1, action.payload.title);

      return adapter.addOne(todo, { ...state, lastId: todo.id });
    }

    case TodosActionTypes.Toggle: {
      const todo = state.entities[action.payload.id];

      return adapter.updateOne({
        id     : todo.id,
        changes: { completed: !todo.completed },
      }, state);
    }

    default: {
      return state;
    }
  }
}
