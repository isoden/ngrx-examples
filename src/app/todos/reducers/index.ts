import { createFeatureSelector, createSelector } from '@ngrx/store';
import * as fromItems from './items';
import * as fromVisibilityFilter from './visibility-filter';

export const reducers = {
  items           : fromItems.reducer,
  visibilityFilter: fromVisibilityFilter.reducer,
};

const {
  selectAll: selectAllTodos,
} = fromItems.adapter.getSelectors();

export interface TodosState {
  items: fromItems.TodosState;
  visibilityFilter: fromVisibilityFilter.VisibilityFilterState;
}

export interface State {
  todos: TodosState;
}

export const selectTodos = createFeatureSelector<TodosState>('todos');
export const selectTodosItems = createSelector(selectTodos, state => state.items);
export const selectTodosVisibilityFilter = createSelector(selectTodos, state => state.visibilityFilter);
export const getVisibilityFilter = createSelector(selectTodosVisibilityFilter, fromVisibilityFilter.getVisibilityFilter);
export const getTodos = createSelector(selectTodosItems, selectAllTodos);
export const getFilteredTodos = createSelector(getVisibilityFilter, getTodos, (filter, todos) => {
  if (filter === 'SHOW_ALL') {
    return todos;
  }

  return todos.filter(todo => {
    switch (filter) {
      case 'SHOW_ACTIVE': return !todo.completed;
      case 'SHOW_COMPLETED': return todo.completed;
    }
  });
});
