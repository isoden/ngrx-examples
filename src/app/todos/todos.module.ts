import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { RouterModule } from '@angular/router';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { TodosPageComponent } from './containers/todos-page/todos-page.component';
import { TodoInputComponent } from './components/todo-input/todo-input.component';
import { TodoComponent } from './components/todo/todo.component';
import { TodoFilterComponent } from './components/todo-filter/todo-filter.component';
import { StoreModule } from '@ngrx/store';
import { reducers } from './reducers';

@NgModule({
  imports: [
    CommonModule,
    FormsModule,
    ReactiveFormsModule,
    StoreModule.forFeature('todos', reducers),
    RouterModule.forChild([
      {
        path     : '',
        component: TodosPageComponent,
      }
    ])
  ],
  declarations: [
    TodosPageComponent,
    TodoInputComponent,
    TodoComponent,
    TodoFilterComponent,
  ],
})
export class TodosModule { }
