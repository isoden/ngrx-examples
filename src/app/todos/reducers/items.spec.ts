import { reducer } from './items';
import { Add, Toggle } from '../actions/todos';

describe('Items', () => {
  const addTodos = (titles: string[]) => {
    const initAction = {} as any;

    return titles.reduce((state, title) => reducer(state, new Add({ title })), reducer(undefined, initAction));
  };

  it('should return default state', () => {
    const action = {} as any;
    const result = reducer(undefined, action);

    expect(result).toMatchSnapshot();
  });

  describe('Add', () => {
    it('should add new todo', () => {
      const state = addTodos(['Learn Angular', 'Learn Zone.js', 'Learn RxJS']);

      expect(state).toMatchSnapshot();
    });
  });

  describe('Toggle', () => {
    it('should toggle completed of todo', () => {
      const state = addTodos(['Learn Angular', 'Learn Zone.js', 'Learn RxJS']);
      const result = reducer(state, new Toggle({ id: 2 }));

      expect(result).toMatchSnapshot();
    });
  });
});
