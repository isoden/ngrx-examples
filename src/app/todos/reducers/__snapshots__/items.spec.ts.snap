// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Items Add should add new todo 1`] = `
Object {
  "entities": Object {
    "1": Todo {
      "completed": false,
      "id": 1,
      "title": "Learn Angular",
    },
    "2": Todo {
      "completed": false,
      "id": 2,
      "title": "Learn Zone.js",
    },
    "3": Todo {
      "completed": false,
      "id": 3,
      "title": "Learn RxJS",
    },
  },
  "ids": Array [
    1,
    2,
    3,
  ],
  "lastId": 3,
}
`;

exports[`Items Toggle should toggle completed of todo 1`] = `
Object {
  "entities": Object {
    "1": Todo {
      "completed": false,
      "id": 1,
      "title": "Learn Angular",
    },
    "2": Object {
      "completed": true,
      "id": 2,
      "title": "Learn Zone.js",
    },
    "3": Todo {
      "completed": false,
      "id": 3,
      "title": "Learn RxJS",
    },
  },
  "ids": Array [
    1,
    2,
    3,
  ],
  "lastId": 3,
}
`;

exports[`Items should return default state 1`] = `
Object {
  "entities": Object {},
  "ids": Array [],
  "lastId": 0,
}
`;
